server/src/main/java/org/elasticsearch/cluster/coordination/MutableMasterHistory.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * This class represents a node's view of the history of which nodes have been elected master over the last 30 minutes. It is kept in
+ * memory, so when a node comes up it does not have any knowledge of previous master history before that point. This object is updated
+ * if and when the cluster state changes with a new master node.
+ */
+public class MutableMasterHistory implements MasterHistory, ClusterStateListener {
```

Comment:
The Mutable/Immutable naming is a bit confusing to me, although I understand where you're coming from. 

I believe the need for such distinction arises from the fact that the entity carries logic with it?

Could having the `Response` for the master history action encapsulate a `List<TimeAndMaster>` and send that over the wire be enough? We could have the `getMostRecentNonNullMaster`, `hasSameMasterGoneNullNTimes` and `getDistinctMasterSeen` (which are kind of duplicated on the `ImmutableMasterHistory`) be static methods (on the MasterHistoryService?) that receive the `List<TimeANdMaster>` list as parameter.

Would this be enough to eliminate the need to differentiate mutable/immutable entities?

Reply:
I changed it to List<DiscoveryNode>. It simplifies serialization a little, but makes using it a little more awkward. I'm on the fence about this change.